docs/dev/topology-over-raft.md

```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+- Immediately reject the joining node if it is able to verify some parameters
+  on its own and tell that they are incorrect.
+
+In case of other failures (timeouts, connection issues etc.) the request is
+simply retried.
+
+After placing the join request, the node needs to wait until it receives and
+processes a reply. Then, it can start its group 0 server and continue with
+boot.
+
+On receiving `JOIN_NODE_REPLY`, the node first verifies if the cookie included
+in reply matches the cookie that is currently loaded to memory - if it doesn't,
+the response should be rejected with appropriate error variant. Then it should
+process the coordinator's decision:
+- If the new node was rejected, the node should fail the startup procedure.
+- If the new node was accepted:
+  1. The joining node must persist enough information so that it won't change
+     its current parameters in an unsafe way, e.g. features enabled
+     by the cluster must be persisted.
+  2. Features currently enabled by the cluster must be enabled locally.
+  3. The join cookie must be persisted in the system keyspace.
+  4. The node responds with a success.
+
+*Request recipient node behavior*
+
+On receiving `JOIN_NODE_REQUEST`, the node verifies some parameters included
+in request. If the verification fails, a response is immediately sent with
+information about rejection. If the verification is successful, the node
+looks at the `system.topology` under a group 0 guard:
+- If there is already an existing request with the same join cookie, an `ok`
+  response is sent back to the joining node, telling it to wait for a response
+  from the coordinator.
+- If there is already an existing request with a different join cookie, the
+  node is told to retry later after the coordinator notices this request and
+  eventually removes it.
+- If there is no topology request for that node, it is placed in
+  the `system.topology` table and an `ok` response is sent back.
+- If the node has already joined, an `ok` response is sent back.
+
+*Topology coordinator behavior*
+
+Upon noticing a node in `none` state and with `join` topology request, the
+coordinator verifies parameters of the node (e.g. whether it supports all
+of the enabled features) and decides whether to accept it or not. It sends
+a `JOIN_NODE_RESPONSE` RPC to the node, telling it about the decision.
+- If the joining node responds with `old_cookie`, the request is simply
+  removed from the `system.topology` table.
+- If the joining node accepts the decision and responds successfully:
+  - If the decision was to add the node, then it is added to raft group 0 and
+    transitioned to the appropriate state (`bootstrapping` or `replacing`).
+    The join cookie is cleared from the `system.topology` table.
+  - If the decision was to reject the node, then the request is simply removed
+    from the `system.topology` table.
+
+In case of other failures (timeouts, connection issues etc.) the request is
+simply retried.
```

Comment:
If a node that requests joining is permanently dead we need to stop retrying eventually.

Reply:
Yes. This should be implemented as a part of the error handling work.